go-controller/pkg/ovn/controller/services/node_tracker.go

```diff
@@ -147,28 +164,45 @@ func (nt *nodeTracker) updateNodeInfo(nodeName, switchName, routerName string, n
 	}
 
 	nt.nodes[nodeName] = ni
+	if chassisID != "" {
+		// Services are currently supported only on the node's first IP.
+		// Extract that one and populate the node's IP template value.
+		if globalconfig.IPv4Mode {
+			if ipv4, err := util.MatchFirstIPFamily(false, nodeIPs); err == nil {
+				nt.nodeIPv4Template.Value[chassisID] = ipv4.String()
+			}
+		}
+		if globalconfig.IPv6Mode {
+			if ipv6, err := util.MatchFirstIPFamily(true, nodeIPs); err == nil {
+				nt.nodeIPv6Template.Value[chassisID] = ipv6.String()
+			}
+		}
+	}
+	nodeIPv4Template, nodeIPv6Template := *nt.nodeIPv4Template, *nt.nodeIPv6Template
 	nt.Unlock()
 
 	klog.Infof("Node %s switch + router changed, syncing services", nodeName)
 	// Resync all services
-	nt.resyncFn()
-}
-
-// removeNodeWithServiceReSync removes a node from the LB -> node mapper
-// *and* forces full reconciliation of services.
-func (nt *nodeTracker) removeNodeWithServiceReSync(nodeName string) {
-	nt.removeNode(nodeName)
-	nt.resyncFn()
+	nt.resyncFn(&nodeIPv4Template, &nodeIPv6Template)
 }
 
 // RemoveNode removes a node from the LB -> node mapper
-// We don't need to re-sync here, because any stale LBs
-// will eventually be cleaned up, and they don't have any cost.
-func (nt *nodeTracker) removeNode(nodeName string) {
+// If 'doResync' is true then this also forces a full reconciliation of
+// services.
+func (nt *nodeTracker) removeNode(nodeName string, doResync bool) {
 	nt.Lock()
-	defer nt.Unlock()
 
+	if node, found := nt.nodes[nodeName]; found {
+		delete(nt.nodeIPv4Template.Value, node.chassisID)
+		delete(nt.nodeIPv6Template.Value, node.chassisID)
+	}
 	delete(nt.nodes, nodeName)
+	nodeIPv4Template, nodeIPv6Template := *nt.nodeIPv4Template, *nt.nodeIPv6Template
```

Comment:
This is not really a copy and can still change. Should we copy?

Reply:
Yes, we should, I'm not sure what I was thinking when I wrote this.  I'll take care of it, thanks for spotting it!